include/sparrow/utils/metadata.hpp

```diff
@@ -0,0 +1,196 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <concepts>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <numeric>
+#include <ranges>
+#include <string_view>
+#include <utility>
+
+#include "sparrow/config/config.hpp"
+#include "sparrow/utils/contracts.hpp"
+#include "sparrow/utils/pair.hpp"
+
+
+#if defined(__cpp_lib_format)
+#    include <format>
+#    include <ostream>
+#endif
+
+namespace sparrow
+{
+    using metadata_key = std::string_view;
+    using metadata_value = std::string_view;
+    using metadata_pair = std::pair<metadata_key, metadata_value>;
+
+    // Helper function to extract an int32 from a char buffer
+    SPARROW_API int32_t extract_int32(const char*& ptr);
+
+    class key_value_view;
+
+    class key_value_view_iterator
+    {
+    public:
+
+        using iterator_category = std::input_iterator_tag;
+        using value_type = metadata_pair;
+        using difference_type = std::ptrdiff_t;
+        using pointer = value_type*;
+        using reference = value_type&;
+
+        SPARROW_API key_value_view_iterator(const key_value_view& parent, int32_t index);
+
+        SPARROW_API value_type operator*() const;
+
+        SPARROW_API key_value_view_iterator& operator++();
+
+        friend bool operator==(const key_value_view_iterator& lhs, const key_value_view_iterator& rhs)
+        {
+            return lhs.m_index == rhs.m_index;
+        }
+
+        friend bool operator!=(const key_value_view_iterator& lhs, const key_value_view_iterator& rhs)
+        {
+            return !(lhs == rhs);
+        }
+
+    private:
+
+        SPARROW_API std::string_view extract_string_view();
+        SPARROW_API void extract_key_value();
+
+        const key_value_view& m_parent;
+        int32_t m_index;
+        const char* m_current;
+        std::string_view m_key;
+        std::string_view m_value;
+    };
+
+    // Custom view to lazily extract key/value pairs from the buffer
+    class key_value_view : public std::ranges::view_interface<key_value_view>
+    {
+    public:
+
+        SPARROW_API key_value_view(const char* ptr);
+
+        [[nodiscard]] SPARROW_API key_value_view_iterator cbegin() const;
+
+        [[nodiscard]] SPARROW_API key_value_view_iterator begin() const;
+
+        [[nodiscard]] SPARROW_API key_value_view_iterator cend() const;
+
+        [[nodiscard]] SPARROW_API key_value_view_iterator end() const;
+
+        [[nodiscard]] SPARROW_API size_t size() const;
+
+    private:
+
+        const char* m_ptr;
+        int32_t m_num_pairs = 0;
+
+        friend key_value_view_iterator;
+    };
+
+    template <typename T>
+    concept input_metadata_container = std::ranges::input_range<T>
+                                       && std::same_as<std::ranges::range_value_t<T>, metadata_pair>;
+
+    std::string get_metadata_from_key_values(const input_metadata_container auto& metadata)
```

Comment:
Styling: an template method would be more consistent with the existing code base rather than an auto function.

Reply:
ok